src/aliased_buffer.h

```diff
@@ -0,0 +1,193 @@
+   * native buffer.
+   *
+   *  Note that byte_offset must by word-aligned.
+   */
+  AliasedBuffer(
+    v8::Isolate* isolate,
+    const size_t byte_offset,
+    const size_t count,
+    const AliasedBuffer<uint8_t,
+    v8::Uint8Array>& backingBuffer) :
+    isolate_(isolate),
+    count_(count),
+    byte_offset_(byte_offset),
+    freeBuffer_(false) {
+    const v8::HandleScope handle_scope(this->isolate_);
+
+    this->buffer_ = reinterpret_cast<NativeT*>(
+      const_cast<uint8_t*>(backingBuffer.GetNativeBuffer() + byte_offset));
+    v8::Local<v8::ArrayBuffer> ab = backingBuffer.GetArrayBuffer();
+    v8::Local<V8T> jsArray = V8T::New(ab, byte_offset, count);
+    this->jsArray_ = v8::Global<V8T>(isolate, jsArray);
+  }
+
+  AliasedBuffer(const AliasedBuffer& that) :
+    isolate_(that.isolate_),
+    count_(that.count_),
+    byte_offset_(that.byte_offset_),
+    buffer_(that.buffer_),
+    freeBuffer_(false) {
+    this->jsArray_ = v8::Global<V8T>(that.isolate_, that.GetJSArray());
+  }
+
+  ~AliasedBuffer() {
+    if (this->freeBuffer_ && this->buffer_ != NULL) {
+      free(this->buffer_);
+    }
+    // TODO(@mike-kaufman):  what do we do about freeing the
+    // v8::Global<ArrayBuffer> & the v8::Global<U>
+  }
+
+  /**
+   * Helper class that is returned from operator[] to support assignment into
+   * a specified location.
+   */
+  class Reference {
+   public:
+    Reference(AliasedBuffer<NativeT, V8T>* aliasedBuffer, size_t index) :
+      aliasedBuffer_(aliasedBuffer),
+      index_(index) {
+    }
+
+    Reference(const Reference& that) :
+      aliasedBuffer_(that.aliasedBuffer_),
+      index_(that.index_) {
+    }
+
+    inline Reference& operator=(const NativeT &val) {
+      this->aliasedBuffer_->SetValue(this->index_, val);
+      return *this;
+    }
+
+    operator NativeT() {
+      return this->aliasedBuffer_->GetValue(this->index_);
+    }
+
+   private:
+    AliasedBuffer<NativeT, V8T>* aliasedBuffer_;
+    size_t index_;
+  };
+
+  /**
+   *  Get the underlying v8 TypedArray overlayed on top of the native buffer
+   */
+  v8::Local<V8T> GetJSArray() const {
+    return this->jsArray_.Get(this->isolate_);
+  }
+
+  /**
+  *  Get the underlying v8::ArrayBuffer underlying the TypedArray and
+  *  overlaying the native buffer
+  */
+  v8::Local<v8::ArrayBuffer> GetArrayBuffer() const {
+    return this->GetJSArray()->Buffer();
+  }
+
+  /**
+   *  Get the underlying native buffer. Note that all reads/writes should occur
+   *  through the GetValue/SetValue/operator[] methods
+   */
+  inline const NativeT* GetNativeBuffer() const {
+    return this->buffer_;
+  }
+
+  /**
+   *  Synonym for GetBuffer()
+   */
+  inline const NativeT* operator * () {
+    return this->GetNativeBuffer();
+  }
+
+  /**
+   *  Set position index to given value.
+   */
+  inline void SetValue(const int index, NativeT value) {
+    // if (index >= count_) {
+    // TODO(@mike-kaufman): Assert?  Fail fast?
```

Comment:
Wasn't clear if I should do anything here.  Please let me know if there's existing precedent around such checks.

Reply:
`CHECK` is likely appropriate. /cc @bnoordhuis

Reply:
added a CHECK.